File: tests/test_manifest.py
"""The manifest template, the compose file and the policy set.

These are the files the ladder and the certifier read, and every assertion
here corresponds to a way a Coworld silently fails: no `num_agents` and the
ladder schedules zero episodes; a pod replay viewer and every hosted replay
hangs; URI-form docs and the coworld page is empty; one protocol instead of
two and the manifest is rejected.
"""

from __future__ import annotations

import json
import os
import re
import stat

from conftest import REPO_ROOT

from cogball import defaults

MANIFEST = json.loads(
    (REPO_ROOT / "coworld_manifest_template.json").read_text())
GAME = MANIFEST["game"]
IMAGE_NAME = "coworld-cogball"
SEATS = 2


def test_num_agents_is_declared_in_every_variant_and_the_fixture():
    assert MANIFEST["variants"], "no variants declared"
    for variant in MANIFEST["variants"]:
        config = variant["game_config"]
        assert config.get("num_agents") == SEATS, variant["id"]
        assert len(config["players"]) == SEATS, variant["id"]
    cert = MANIFEST["certification"]
    assert cert["game_config"]["num_agents"] == SEATS
    assert len(cert["players"]) == SEATS
    assert len(cert["game_config"]["players"]) == SEATS


def test_results_schema_matches_the_servers_closed_key_set():
    from cogball.config import GameConfig
    from cogball.engine import EpisodeResult
    from cogball.server import results_document

    config = GameConfig.from_dict({
        "players": [{"name": "a"}, {"name": "b"}],
        "tokens": ["t0", "t1"], "seed": 1})
    result = EpisodeResult(
        reason="complete", end_rule="full_time", winner=0,
        scores=(1.0, 0.0), win=(True, False), goals=(3, 0),
        final_tick=7200, final_turn=48, shots=(1, 1),
        shots_on_target=(1, 0), saves=(0, 1), passes_completed=(2, 0),
        interceptions=(0, 2), possession_ticks=(1, 1),
        distance_m=(0.0,) * defaults.NUM_ROBOTS, llm_turns=(48, 0),
        fallback_turns=(0, 48),
        fallback_causes=(dict.fromkeys(defaults.FALLBACK_CAUSES, 0),) * 2,
        policy_kinds=("llm", "scripted"))
    produced = set(results_document(config, result))
    schema = GAME["results_schema"]
    assert produced == set(schema["properties"]), (
        "results.json and the manifest results_schema disagree; they are a "
        "CLOSED schema and must be edited together")
    assert produced == set(schema["required"])
    assert schema["additionalProperties"] is False


def test_the_reason_and_end_rule_enums_are_the_closed_ones():
    props = GAME["results_schema"]["properties"]
    assert props["reason"]["enum"] == list(defaults.REASONS)
    assert props["end_rule"]["enum"] == list(defaults.END_RULES)
    causes = props["fallback_causes"]["items"]["properties"]
    assert set(causes) == set(defaults.FALLBACK_CAUSES)


def test_the_replay_viewer_is_the_static_bundle_never_a_pod():
    assert GAME["replay_viewer"] == {"bundle": "static-replay-viewer"}


def test_both_protocols_are_declared():
    protocols = GAME["protocols"]
    assert set(protocols) == {"player", "global"}
    for entry in protocols.values():
        assert entry["type"] == "uri"
        assert entry["value"].startswith(
            "https://github.com/Metta-AI/cogame-cogball/")


def test_docs_are_inline_text_and_track_the_files_they_came_from():
    docs = GAME["docs"]
    readme = docs["readme"]
    assert readme["type"] == "text"
    assert len(readme["value"]) > 1000
    assert readme["value"].splitlines()[0] == \
        (REPO_ROOT / "README.md").read_text().splitlines()[0]
    pages = docs["pages"]
    assert len(pages) == 2
    sources = {"protocol.md": REPO_ROOT / "docs" / "PROTOCOL.md",
               "coaching.md": REPO_ROOT / "docs" / "COACHING.md"}
    for page in pages:
        assert page["content"]["type"] == "text"
        assert len(page["content"]["value"]) > 1000
        assert page["title"]
        assert page["content"]["value"].splitlines()[0] == \
            sources[page["id"]].read_text().splitlines()[0], (
                f"{page['id']} has drifted from its source file; re-inline it")


def test_the_episode_budget_leaves_room_for_the_engine_to_settle():
    assert MANIFEST["episode_timeout_minutes"] == 20
    platform_seconds = MANIFEST["episode_timeout_minutes"] * 60
    for variant in MANIFEST["variants"]:
        budget = variant["game_config"]["wall_clock_budget_seconds"]
        assert budget <= 0.6 * platform_seconds, variant["id"]
    assert defaults.PLATFORM_EPISODE_TIMEOUT_MINUTES == \
        MANIFEST["episode_timeout_minutes"]
    assert defaults.DEFAULT_WALL_CLOCK_BUDGET_SECONDS == \
        MANIFEST["variants"][0]["game_config"]["wall_clock_budget_seconds"]


def test_the_certification_fixture_is_offline_and_quick():
    cert = MANIFEST["certification"]
    assert [p["player_id"] for p in cert["players"]] == ["baseline"] * SEATS
    baseline = next(p for p in MANIFEST["player"] if p["id"] == "baseline")
    assert baseline["env"] == {"PLAYER_SCRIPTED": "formation"}
    assert baseline["run"] == ["/bin/cogball-player"]
    assert "PLAYER_PROMPT" not in baseline.get("env", {})
    assert cert["game_config"]["max_ticks"] == 900


def test_the_game_entrypoint_and_source_url():
    assert GAME["name"] == "cogball"
    assert GAME["runnable"]["run"] == ["/bin/cogball"]
    assert GAME["runnable"]["image"] == "{{GAME_IMAGE}}"
    assert GAME["runnable"]["source_url"].endswith("cogame-cogball/tree/main")


def test_compose_uses_one_image_and_it_matches_the_scaffold():
    compose = (REPO_ROOT / "compose.yaml").read_text()
    images = re.findall(r"^\s+image:\s*(\S+)\s*$", compose, re.M)
    assert images == [f"{IMAGE_NAME}:latest"] * 2, images
    assert compose.count("platform: linux/amd64") == 2
    ci = (REPO_ROOT / ".github" / "workflows" / "ci.yml").read_text()
    assert f"IMAGE: {IMAGE_NAME}" in ci


def test_the_policy_set_is_two_prompts_two_baselines_one_image():
    policies = json.loads((REPO_ROOT / "tools" / "ci" / "policies.json").read_text())
    assert len(policies) == 4
    names = [p["name"] for p in policies]
    assert names == ["cogball-total", "cogball-counter",
                     "cogball-formation", "cogball-swarm"]
    for policy in policies:
        assert policy["run"] == "/bin/cogball-player"
        assert "image" not in policy      # one image, env-switched
    prompts = [p for p in policies if "PLAYER_PROMPT" in p["env"]]
    scripted = [p for p in policies if "PLAYER_SCRIPTED" in p["env"]]
    assert len(prompts) == 2 and len(scripted) == 2
    # the two champions must be genuinely different policies, or the second
    # upload dedupes onto the first version
    assert prompts[0]["env"]["PLAYER_PROMPT"] != prompts[1]["env"]["PLAYER_PROMPT"]
    assert "player" not in prompts[0]
    assert prompts[1]["player"] == "ply_bac48eb1-662e-44f8-973d-f3e016dccf5d"
    assert {p["env"]["PLAYER_SCRIPTED"] for p in scripted} == \
        set(defaults.BASELINES)


def test_the_ci_scaffold_has_no_unsubstituted_placeholders():
    files = [".github/workflows/ci.yml",
             ".github/workflows/coworld-release.yml",
             ".github/workflows/coworld-submit.yml",
             "tools/ci/docker_smoke.sh",
             "tools/ci/policies.json"]
    for name in files:
        text = (REPO_ROOT / name).read_text()
        for placeholder in ("<slug>", "<IMAGE>", "<SEATS>"):
            assert placeholder not in text, f"{name} still contains {placeholder}"


def test_the_ci_hooks_are_committed_executable():
    for name in ("tools/ci/docker_smoke.sh", "tools/build_replay_viewer.sh"):
        path = REPO_ROOT / name
        assert path.exists(), name
        assert os.stat(path).st_mode & stat.S_IXUSR, (
            f"{name} must be executable; fix with "
            f"git update-index --chmod=+x {name}")


def test_the_release_and_submit_workflows_expose_the_inputs_later_phases_pass():
    release = (REPO_ROOT / ".github" / "workflows" /
               "coworld-release.yml").read_text()
    for name in ("version:", "policies:", "put_secret:", "skip_certify:"):
        assert name in release
    assert "release-result" in release
    assert '"player"' in release      # per-policy owner for champion #2
    submit = (REPO_ROOT / ".github" / "workflows" /
              "coworld-submit.yml").read_text()
    for name in ("player_id:", "policy:", "league_id:"):
        assert name in submit
    assert "submit-result" in submit

File: tests/test_manifest.nim
## The manifest is a contract with the platform's certifier, and every clause
## in it has a counterpart in the code. This suite is where the two are held
## together: a manifest edited without the sim (or the reverse) fails here
## rather than in hosted certification.

import std/[json, os, strutils]
import lib/helpers


const ManifestPath = "coworld_manifest_template.json"

let manifest = parseJson(readFile(ManifestPath))

proc seatCountEverywhere() =
  ## `num_agents` is 2 in EVERY variant and in the certification fixture, and
  ## every other seat-count declaration agrees with it.
  doAssert manifest.hasKey("variants")
  doAssert manifest["variants"].len >= 2
  for variant in manifest["variants"]:
    let config = variant["game_config"]
    doAssert config.hasKey("num_agents"),
      "variant " & variant["id"].getStr() & " has no num_agents"
    doAssert config["num_agents"].getInt == SeatCount,
      "variant " & variant["id"].getStr() & " seats " &
        $config["num_agents"].getInt
    doAssert config["players"].len == SeatCount
    doAssert config["slots"].len == SeatCount
    doAssert config["minPlayers"].getInt == SeatCount
  let cert = manifest["certification"]
  doAssert cert["game_config"]["num_agents"].getInt == SeatCount
  doAssert cert["players"].len == SeatCount
  doAssert cert["game_config"]["players"].len == SeatCount
  report "num_agents is " & $SeatCount &
    " in every variant and the cert fixture"

proc budgetIsInsideTheEpisodeTimeout() =
  ## 60 % of the platform's 1200 s episode timeout is the design's ceiling.
  const EpisodeTimeoutSeconds = 1200
  for variant in manifest["variants"]:
    let config = variant["game_config"]
    let budget = config["wallClockBudgetSeconds"].getInt
    doAssert budget * 10 <= EpisodeTimeoutSeconds * 6,
      "variant " & variant["id"].getStr() & " budgets " & $budget &
        " s against a " & $EpisodeTimeoutSeconds & " s timeout"
    # The turn budget has to cover both attempts.
    doAssert config["turnBudgetMs"].getInt >=
      config["attempt1Ms"].getInt + config["retryMs"].getInt
    # And the turns have to fit inside the budget with room for the match.
    let turns = config["maxTicks"].getInt div config["turnTicks"].getInt
    doAssert turns * config["turnBudgetMs"].getInt div 1000 < budget,
      "variant " & variant["id"].getStr() & ": " & $turns &
        " turns cannot fit in " & $budget & " s"
  report "every variant settles well inside 60% of the episode timeout"

proc resultsSchemaMatchesTheCode() =
  ## The schema is `additionalProperties: false`, so a key the code writes and
  ## the schema does not declare is a rejected episode.
  var sim = playing(testConfig())
  sim.finishGame(reasonComplete, erFullTime)
  let produced = parseJson(sim.playerResultsJson())
  let declared = manifest["game"]["results_schema"]["properties"]
  doAssert declared.len == produced.len,
    "the schema declares " & $declared.len & " keys, the code writes " &
      $produced.len
  for key, _ in produced:
    doAssert declared.hasKey(key),
      "playerResultsJson writes `" & key & "`, the schema does not declare it"
  for key, _ in declared:
    doAssert produced.hasKey(key),
      "the schema declares `" & key & "`, playerResultsJson never writes it"
  doAssert manifest["game"]["results_schema"]["additionalProperties"].getBool ==
    false
  for key in ["names", "scores", "win", "team", "goals", "reason", "endRule"]:
    var required = false
    for entry in manifest["game"]["results_schema"]["required"]:
      if entry.getStr() == key:
        required = true
    doAssert required, key & " must be required"
  let reasons = manifest["game"]["results_schema"]["properties"]["reason"]["enum"]
  doAssert reasons.len == 3
  for reason in EndReason:
    var present = false
    for entry in reasons:
      if entry.getStr() == reasonText(reason):
        present = true
    doAssert present, "the reason enum is missing " & reasonText(reason)
  let rules = manifest["game"]["results_schema"]["properties"]["endRule"]["enum"]
  for rule in EndRule:
    var present = false
    for entry in rules:
      if entry.getStr() == endRuleText(rule):
        present = true
    doAssert present, "the endRule enum is missing " & endRuleText(rule)
  for key, spec in declared:
    if spec.hasKey("items"):
      doAssert spec["minItems"].getInt == 2 and spec["maxItems"].getInt == 2,
        key & " is not pinned to two seats"
  report "results_schema matches playerResultsJson key for key"

proc configSchemaCoversWhatTheCodeReads() =
  ## Every field `sim_config.update` reads must be declared, or a variant that
  ## sets it is rejected by the certifier.
  let declared = manifest["game"]["config_schema"]["properties"]
  let source = readFile("src/cogball/sim_config.nim")
  var missing: seq[string]
  for line in source.splitLines():
    let trimmed = line.strip()
    for reader in ["node.readInt(\"", "node.readBool(\"", "node.readStr(\""]:
      if trimmed.startsWith(reader):
        let rest = trimmed[reader.len .. ^1]
        let key = rest[0 ..< rest.find('"')]
        if not declared.hasKey(key) and key notin ["numAgents"]:
          missing.add(key)
  doAssert missing.len == 0,
    "config_schema does not declare: " & missing.join(", ")
  for key in ["tokens", "players", "slots"]:
    doAssert declared.hasKey(key), "config_schema lacks " & key
  doAssert manifest["game"]["config_schema"]["additionalProperties"].getBool ==
    false
  var required: seq[string]
  for entry in manifest["game"]["config_schema"]["required"]:
    required.add(entry.getStr())
  doAssert required == @["tokens", "players"]
  report "config_schema covers every field sim_config.update reads"

proc viewerIsAStaticBundle() =
  doAssert manifest["game"]["replay_viewer"]["bundle"].getStr() ==
    "static-replay-viewer",
    "the replay viewer must be the STATIC bundle, never a pod"
  report "replay_viewer.bundle is static-replay-viewer"

proc protocolsAndDocs() =
  let protocols = manifest["game"]["protocols"]
  for side in ["player", "global"]:
    doAssert protocols.hasKey(side), "game.protocols lacks " & side
    doAssert protocols[side]["value"].getStr().len > 0
  let docs = manifest["game"]["docs"]
  doAssert docs["readme"]["type"].getStr() == "text",
    "game.docs.readme must be inline TEXT, not a URI"
  doAssert docs["readme"]["value"].getStr().len > 1000,
    "the inlined readme is suspiciously short"
  doAssert docs["pages"].len == 3
  var ids: seq[string]
  for page in docs["pages"]:
    ids.add(page["id"].getStr())
    doAssert page["title"].getStr().len > 0
    doAssert page["content"]["type"].getStr() == "text"
    doAssert page["content"]["value"].getStr().len > 1000,
      "page " & page["id"].getStr() & " is suspiciously short"
  doAssert ids == @["rules.md", "protocol.md", "coaching.md"], $ids
  # And the inlined text is the CURRENT docs, not a stale paste.
  doAssert docs["readme"]["value"].getStr() == readFile("README.md")
  doAssert docs["pages"][0]["content"]["value"].getStr() ==
    readFile("docs/RULES.md")
  doAssert docs["pages"][1]["content"]["value"].getStr() ==
    readFile("docs/PROTOCOL.md")
  doAssert docs["pages"][2]["content"]["value"].getStr() ==
    readFile("docs/COACHING.md")
  report "both protocols are declared and all four docs are inline, current text"

proc imageAndEntrypoints() =
  let compose = readFile("compose.yaml")
  doAssert compose.contains("image: coworld-cogball:latest")
  doAssert compose.contains("platform: linux/amd64")
  doAssert compose.contains("network: host")
  doAssert compose.contains("cogball:"), "the compose SERVICE is not named for the coworld"
  doAssert manifest["game"]["runnable"]["image"].getStr() == "{{COGBALL_IMAGE}}",
    "the manifest placeholder does not match the compose service name"
  doAssert manifest["game"]["runnable"]["run"][0].getStr() == "/bin/cogball"
  doAssert manifest["game"]["runnable"]["env"]["ANTHROPIC_API_KEY_URI"]
    .getStr() == "secret://coworld/cogball/anthropic_api_key"
  doAssert manifest["player"][0]["image"].getStr() == "{{COGBALL_IMAGE}}",
    "the bundled player must come out of the SAME image"
  doAssert manifest["player"][0]["run"][0].getStr() == "/bin/cogball-player"
  doAssert manifest["player"][0]["env"]["PLAYER_SCRIPTED"].getStr() ==
    "formation", "the certification player must be the scripted baseline"
  let dockerfile = readFile("Dockerfile")
  doAssert dockerfile.contains("/bin/cogball") and
    dockerfile.contains("/bin/cogball-player"),
    "the image does not ship both entrypoints"
  report "one image, two entrypoints, and the placeholder matches compose"

proc policiesAreTheRightShape() =
  ## Two LLM prompt policies and two scripted baselines, all env-switched out
  ## of one image, with champion #2 owned by the second identity.
  let policies = parseJson(readFile("tools/ci/policies.json"))
  doAssert policies.len == 4, "expected four policies, saw " & $policies.len
  var prompts = 0
  var scripted = 0
  var owned = 0
  var names: seq[string]
  for policy in policies:
    names.add(policy["name"].getStr())
    doAssert policy["run"].getStr() == "/bin/cogball-player",
      "every policy runs the one player entrypoint"
    if policy["env"].hasKey("PLAYER_PROMPT"):
      inc prompts
      doAssert policy["env"]["PLAYER_PROMPT"].getStr().len > 300,
        policy["name"].getStr() & "'s prompt is too thin to be a strategy"
    if policy["env"].hasKey("PLAYER_SCRIPTED"):
      inc scripted
      doAssert policy["env"]["PLAYER_SCRIPTED"].getStr() in
        ["formation", "swarm"]
    if policy.hasKey("player"):
      inc owned
      doAssert policy["player"].getStr().startsWith("ply_")
  doAssert prompts == 2, "expected two LLM champions"
  doAssert scripted == 2, "expected two scripted fillers"
  doAssert owned == 1, "champion #2 must carry its owning player id"
  doAssert names == @["cogball-total", "cogball-counter", "cogball-formation",
    "cogball-swarm"], $names
  # The two champion prompts must actually differ.
  doAssert policies[0]["env"]["PLAYER_PROMPT"].getStr() !=
    policies[1]["env"]["PLAYER_PROMPT"].getStr()
  report "policies.json: two prompts, two baselines, one image, owner pinned"

proc scaffoldIsExecutable() =
  for path in ["tools/ci/docker_smoke.sh", "tools/build_replay_viewer.sh"]:
    doAssert fileExists(path), path & " is missing"
    let permissions = getFilePermissions(path)
    doAssert fpUserExec in permissions,
      path & " must be committed executable (mode 100755)"
  for path in [".github/workflows/ci.yml",
               ".github/workflows/coworld-release.yml",
               ".github/workflows/coworld-submit.yml",
               "tools/ci/policies.json"]:
    doAssert fileExists(path), path & " is missing"
  report "the CI scaffold is present and both hooks are executable"

proc noUnsubstitutedPlaceholders() =
  ## The three scaffold placeholders, and only those three. The four
  ## angle-bracket names that survive by design are runtime values.
  const Expected = ["<run_id>", "<name>", "<cow_id>", "<sha>"]
  for path in [".github/workflows/ci.yml",
               ".github/workflows/coworld-release.yml",
               ".github/workflows/coworld-submit.yml",
               "tools/ci/docker_smoke.sh", "tools/ci/policies.json"]:
    let text = readFile(path)
    for placeholder in ["<slug>", "<IMAGE>", "<SEATS>"]:
      doAssert not text.contains(placeholder),
        path & " still carries " & placeholder
    var i = 0
    while i < text.len:
      let open = text.find('<', i)
      if open < 0:
        break
      let close = text.find('>', open)
      if close < 0 or close - open > 12:
        i = open + 1
        continue
      let token = text[open .. close]
      var ok = true
      for ch in token[1 ..< token.len - 1]:
        if ch notin {'A' .. 'Z', 'a' .. 'z', '0' .. '9', '_'}:
          ok = false
      if ok and token notin Expected:
        doAssert false, path & " carries an unexpected placeholder " & token
      i = open + 1
  report "no unsubstituted placeholder survives in the scaffold"

when isMainModule:
  echo "test_manifest"
  seatCountEverywhere()
  budgetIsInsideTheEpisodeTimeout()
  resultsSchemaMatchesTheCode()
  configSchemaCoversWhatTheCodeReads()
  viewerIsAStaticBundle()
  protocolsAndDocs()
  imageAndEntrypoints()
  policiesAreTheRightShape()
  scaffoldIsExecutable()
  noUnsubstitutedPlaceholders()
  echo "test_manifest: all good"
